src/main/scala/io/iohk/ethereum/blockchain/sync/SyncStateSchedulerActor.scala

```diff
@@ -97,73 +167,298 @@ class SyncStateSchedulerActor(downloader: ActorRef, sync: SyncStateScheduler, sy
+                case Right((newState, stats)) =>
+                  ProcessingResult(
+                    Right(ProcessingSuccess(newState, newDownloaderState, currentStats.addStats(stats)))
+                  )
+              }
+          }
+        case RequestFailed(from, reason) =>
+          log.debug("Processing failed request from {}. Failure reason {}", from, reason)
+          val newDownloaderState = currentDownloaderState.handleRequestFailure(from)
+          ProcessingResult(Left(DownloaderError(newDownloaderState, from, reason, critical = true)))
+      }
+    }
+  }
+
+  private def handleRestart(
+      currentState: SchedulerState,
+      currentStats: ProcessingStatistics,
+      targetBlock: BigInt,
+      restartRequester: ActorRef
+  ): Unit = {
+    log.debug("Starting request sequence")
+    sync.persistBatch(currentState, targetBlock)
+    restartRequester ! WaitingForNewTargetBlock
+    context.become(idle(currentStats.addSaved(currentState.memBatch.size)))
+  }
+  import akka.pattern.pipe
+
+  // scalastyle:off
   def syncing(
       currentState: SchedulerState,
+      currentDownloaderState: DownloaderState,
       currentStats: ProcessingStatistics,
       targetBlock: BigInt,
-      syncInitiator: ActorRef
-  ): Receive = {
-    case MissingNodes(nodes, downloaderCap) =>
-      log.debug(s"Received {} new nodes to process", nodes.size)
-      // Current SyncStateDownloaderActor makes sure that there is no not requested or duplicated values in its response.
-      // so we can ignore those errors.
-      //TODO [ETCM-275] process responses asynchronously to avoid steep rise of pending requests after pivot block update
-      sync.processResponses(currentState, nodes) match {
-        case Left(value) =>
-          log.error(s"Critical error while state syncing ${value}, stopping state sync")
-          // TODO we should probably start sync again from new target block, as current trie is malformed or declare
-          // fast sync as failure and start normal sync from scratch
-          context.stop(self)
-        case Right((newState, statistics)) =>
-          if (newState.numberOfPendingRequests == 0) {
-            reportStats(syncInitiator, currentStats, currentState)
-            finalizeSync(newState, targetBlock, syncInitiator)
-          } else {
-            log.debug(
-              s" There are {} pending state requests," +
-                s"Missing queue size is {} elements",
-              newState.numberOfPendingRequests,
-              newState.queue.size()
+      syncInitiator: ActorRef,
+      nodesToProcess: Queue[RequestResult],
+      processing: Boolean,
+      restartRequested: Option[ActorRef]
+  ): Receive = handleCommonMessages orElse handleRequestResults orElse {
+    case Sync if currentState.numberOfPendingRequests > 0 && restartRequested.isEmpty =>
+      val freePeers = getFreePeers(currentDownloaderState)
+      nodesToProcess.dequeueOption match {
```

Comment:
minor: since `freePeers` is always used with emptiness check this could be changed into:
`(nodesToProcess.dequeueOption, NonEmptyList.fromList(freePeers)) match {
//rest of code
`
Nice bonus of that approach:
 - no need to use `fromListUnsafe`
 - exhaustiveness checks will work (using `if` in `case` disabled exhaustiveness checker for given match expression)

Reply:
great suggestion!